### PR TITLE
[plugin-web-app] Add possibility to mock browser's HTTP requests

### LINKED ISSUE #1104: I would like to have the possibility to mock requests via proxy.
I would like to have the possibility to mock requests via proxy.

### PR DESCRIPTION
**Pull request by ikalinin1**
closes #1104

### CODE REVIEW

File: vividus/src/main/java/org/vividus/bdd/steps/BddVariableSteps.java [hunk 1]
```diff
@@ -239,6 +240,26 @@ public void initVariableWithGivenValues(Set<VariableScope> scopes, String variab
...
+     * <ul>
+     * <li><b>STEP</b> - the variable will be available only within the step,
+     * <li><b>SCENARIO</b> - the variable will be available only within the scenario,
+     * <li><b>STORY</b> - the variable will be available within the whole story,
+     * <li><b>NEXT_BATCHES</b> - the variable will be available starting from next batch
+     * </ul>
+     * @param variableName A name of variable to assign the values
+     * @param path to a file
+     */
+    @When("I initialize $scopes variable `$variableName` with content of file from path `$path`")
```
Comment (uarlouski) [inline vividus/src/main/java/org/vividus/bdd/steps/BddVariableSteps.java:257]: maybe 
@When("I initialize $scopes variable `$variableName` with binary content of file from path `$path`")

File: vividus-tests/src/main/resources/story/integration/ProxyStepsTests.story [hunk 2]
```diff
@@ -55,3 +55,47 @@ Then a JSON element from '${response}' by the JSON path '$.headers' is equal to
...
+Given I am on a page with the URL '${vividus-test-site-url}/frames.html'
+When I switch to frame located `<frameId>`
+Then number of elements found by `<elementSelector>` is = `1`
+When I mock HTTP responses with request URL which CONTAINS `example.com` using response code `404` and headers:
+|name          |value|
+|Content-Length|0    |
+When I refresh the page
+When I switch to frame located `<frameId>`
+Then number of elements found by `<elementSelector>` is = `0`
+
```
Comment (valfirst) [inline vividus-tests/src/main/resources/story/integration/ProxyStepsTests.story:98]: remove this line to stick examples to their scenario

**Review by valfirst — CHANGES_REQUESTED** [review]

File: docs/modules/commons/pages/expressions.adoc [hunk 3]
```diff
@@ -193,6 +193,25 @@ Loads the resource found at the provided path and replaces the expression with t
 When I initialize the scenario variable `my-data` with value `#{loadResource(/data/body.txt)}`
 ----
 
+=== `loadResourceAsByteArray`
```
Comment (valfirst) [inline docs/modules/commons/pages/expressions.adoc:196]: `loadBinaryResource` ?

File: docs/modules/commons/pages/expressions.adoc [hunk 4]
```diff
@@ -193,6 +193,25 @@ Loads the resource found at the provided path and replaces the expression with t
...
+
+Loads the resource found at the provided path as bytes array. Could be useful for the steps that accepting raw binary data.
```
Comment (valfirst) [inline docs/modules/commons/pages/expressions.adoc:198]: `as bytes array` -> `as bytes`

File: docs/modules/commons/pages/expressions.adoc [hunk 5]
```diff
@@ -193,6 +193,25 @@ Loads the resource found at the provided path and replaces the expression with t
...
+----
+#{loadResourceAsByteArray(*$pathToResource*)}
+----
+
+* {path-parameter}
+
+.Load data from the project resource as bytes array
+[source,gherkin]
+----
+When I mock HTTP responses with request URL which CONTAINS `frames.html` using response code `200`, content `#{loadResource(page.html)}` and headers:
```
Comment (valfirst) [inline docs/modules/commons/pages/expressions.adoc:210]: the example contains invalid expression reference

File: docs/modules/plugins/pages/plugin-web-app.adoc [hunk 6]
```diff
@@ -588,6 +588,59 @@ Then number of HTTP GET requests with URL pattern `http://httpbin\.org/get` is e
...
+
+{proxy}
+Mocks HTTP response with a provided content
+
+[IMPORTANT]
+No actual request will be executed. Short-circuited response will be returned.
+
+[source,gherkin]
+----
+When I mock http requests with URL $rule `$url` with response code `$code` and content `$content`:$headers
```
Comment (valfirst) [inline docs/modules/plugins/pages/plugin-web-app.adoc:601]: `...I mock HTTP responses with request URL which...`

File: docs/modules/plugins/pages/plugin-web-app.adoc [hunk 7]
```diff
@@ -588,6 +588,59 @@ Then number of HTTP GET requests with URL pattern `http://httpbin\.org/get` is e
...
+
+{proxy}
+Mocks HTTP response
+
+[IMPORTANT]
+No actual request will be executed. Short-circuited response will be returned.
+
+[source,gherkin]
+----
+When I mock http requests with URL $rule `$url` with response code `$code` :$headers
```
Comment (valfirst) [inline docs/modules/plugins/pages/plugin-web-app.adoc:628]: `...I mock HTTP responses with request URL which...`

File: vividus-tests/src/main/resources/properties/configuration.properties [hunk 8]
```diff
@@ -1,3 +1,5 @@
 configuration.profiles=web/headless/chrome
 configuration.suite=integration
 configuration.environments=
+
+bdd.variables.global.vividus-test-site-url=https://vividus-test-site.herokuapp.com
```
Comment (valfirst) [inline vividus-tests/src/main/resources/properties/configuration.properties:5]: `configuration.properties` is not the best place to configure global variables

Comment (valfirst) [reply vividus-tests/src/main/resources/properties/configuration.properties:5]: `vividus-tests/src/main/resources/properties/environment/envinroment.properties` ?

File: vividus-tests/src/main/resources/story/integration/ProxyStepsTests.story [hunk 9]
```diff
@@ -55,3 +55,44 @@ Then a JSON element from '${response}' by the JSON path '$.headers' is equal to
     "Testname2": "testValue2"
 }
 'IGNORING_EXTRA_FIELDS
+
+Scenario: Verify step When I mock http requests with URL $rule `$url` with response code `$code` and content `$content`:$headers mocking resource with bytes
```
Comment (valfirst) [inline vividus-tests/src/main/resources/story/integration/ProxyStepsTests.story:59]: please update the step syntax here and below

**Review by valfirst — CHANGES_REQUESTED** [review]

File: vividus-bdd-engine/src/main/java/org/vividus/bdd/expression/IExpressionProcessor.java [hunk 10]
```diff
@@ -20,5 +20,5 @@
 
 public interface IExpressionProcessor
 {
-    Optional<String> execute(String expression);
+    Optional<Object> execute(String expression);
 }
```
Comment (valfirst) [inline vividus-bdd-engine/src/main/java/org/vividus/bdd/expression/IExpressionProcessor.java:24]:
```java
public interface IExpressionProcessor<T>
{
    Optional<T> execute(String expression);
}
```
?

Comment (valfirst) [reply vividus-bdd-engine/src/main/java/org/vividus/bdd/expression/IExpressionProcessor.java:24]: and in this case if you keep `UnaryExpressionProcessor` you won't need to change each and every expression processor

File: vividus-bdd-engine/src/main/java/org/vividus/bdd/expression/RandomIntExpressionProcessor.java [hunk 11]
```diff
@@ -31,7 +31,7 @@
             "^randomInt\\(" + INT_NUMBER_REGEX + ",\\s*" + INT_NUMBER_REGEX + "\\)$", Pattern.CASE_INSENSITIVE);
 
     @Override
-    public Optional<String> execute(String expression)
+    public Optional<Object> execute(String expression)
```
Comment (valfirst) [inline vividus-bdd-engine/src/main/java/org/vividus/bdd/expression/RandomIntExpressionProcessor.java:34]: `Optional<Integer>` ?

File: vividus-bdd-engine/src/main/java/org/vividus/bdd/expression/FunctionalExpressionProcessor.java [hunk 12]
```diff
@@ -17,25 +17,25 @@
 package org.vividus.bdd.expression;
 
 import java.util.Optional;
-import java.util.function.UnaryOperator;
+import java.util.function.Function;
 import java.util.regex.Matcher;
 import java.util.regex.Pattern;
 
-public class UnaryExpressionProcessor implements IExpressionProcessor
+public class FunctionalExpressionProcessor implements IExpressionProcessor
```
Comment (valfirst) [inline vividus-bdd-engine/src/main/java/org/vividus/bdd/expression/FunctionalExpressionProcessor.java:24]: I would kept `UnaryExpressionProcessor` as a child of `FunctionalExpressionProcessor `

**Review by valfirst — CHANGES_REQUESTED** [review]

**Review by valfirst — APPROVED** [review]

**Review by uarlouski — APPROVED** [review]

**Pull request merged**